numpy/core/src/multiarray/common.h

```diff
@@ -182,6 +183,25 @@ check_and_adjust_axis(int *axis, int ndim)
     return check_and_adjust_axis_msg(axis, ndim, Py_None);
 }
 
+/*
+ * Create a ustring (native string) object from a C string, if one has not
+ * already been created.
+ *
+ * The result is placed into `str`, which should be a pointer to a static
+ * variable.
+ */
+static NPY_INLINE void
+npy_cached_ustring(PyObject **str, char const *cstr)
+{
+    if (*str != NULL) {
+        return;
+    }
+    *str = PyUString_FromString(cstr);
```

Comment:
Maybe `PyUString_InternFromString`? That is what is used inside `multiarraymodule.c`, `intern_strings`. Indeed, possibly just add `axis1` and `axis2` to it? Though overall I think it stil does make sense to keep the ability to pass on a regular `char`, and just ensure it gets translated only when the error occurs.

Reply:
> Though overall I think it stil does make sense to keep the ability to pass on a regular char, and just ensure it gets translated only when the error occurs.

I have a stash that does it that way, and just has a `PyObject *py_msg, char const* c_msg` signature, but it felt messier.

Good spot on `PyUString_InternFromString` - that seems like a reasonable thing to do, although the `um` in those variables seems to stand for `umath`, which would be misleading for interned strings used elsewhere.

Reply:
Not in `multiarraymodule.c` - those have `npy_ma_str...`